### DISCUSSION TITLE
Using Arduino Uno/Nano/Leonardo/Micro to function as programmer/debugger

### DISCUSSION THREAD

**Discussion by mcuee** [avrdudes/avrdude#1088] [Ideas]
Uno and Nano are kind of very popular and the clones (usually with CH340G USB to TTL chip) are pretty cheap. Other than ATxmega (PDI), it seems to cover most of the AVR chips.

They can be loaded with different FW to function as different programmers.

1) stk500v1: Arduino as ISP
https://docs.arduino.cc/built-in-examples/arduino-isp/ArduinoISP

2) stk500v2: ScratchMonkey
https://microtherion.github.io/ScratchMonkey/GettingStarted.html
https://github.com/microtherion/ScratchMonkey

3) AVR910:
* https://github.com/avrdudes/avrdude/issues/1060#issuecomment-1227869284

4) jtag2updi
https://github.com/ElTangas/jtag2updi

5) TPI
https://github.com/pduke/UNO_TPI
https://github.com/james-tate/Arduino-TPI-Programmer

6) DebugWire
https://github.com/felias-fogg/dw-link

5) Enhanced jtag2updi
@askn37's improved jtag2updi firmware
https://github.com/askn37/jtag2updi/

**Comment by mcuee** [comment]
Updated enchmark here.
* https://github.com/avrdudes/avrdude/issues/1060#issuecomment-1228495618

Comparison of Uno based programmers using the Uno ATmega328P as a target and same 32KB hex file.

Interestingly ScratchMonkey is faster than AVR910 in terms of flash writing but not flash verification. `Arduino as ISP` by default only works for 19200 baud so it is always the slowest. But it can be optimized to support 11520bps by commenting out the LED related codes.

| Programmer  | Flash writing | Flash verification |
| ------------- | ------------- |------------- |
| Uno AVR910 500000 bps | 20.11 s | 6.65 s |
| ScratchMonkey stk500v2 | 14.59 s | 14.34 s  |
| AVR ISP shield stk500v1 19200bps | 36.75 s |  20.47 s |
| AVR ISP shield stk500v1 115200bps | 21.99 s |  11.51 s |

**Reply by Ho-Ro** [reply]
#### nanoSTK
See also the [results](https://github.com/avrdudes/avrdude/discussions/1022#discussioncomment-7352277) for [nanoSTK](https://github.com/Ho-Ro/nanoSTK_V1), a highly optimised programmer based on *Arduino Nano* with `stk500v1` protocol.
With a minimal [HW mod of the nano](https://github.com/Ho-Ro/nanoSTK_V1#hw-modifications-for-nanostk) you'll get a programmer well suited for targets equipped with an ISP connector. The modified HW is also compatible with the [ScratchMonkey FW](https://github.com/microtherion/ScratchMonkey).

HW / SW Features:
- Automatic detection of `stk500v1` or `arduino` protocol and adapting the EEPROM address mode (byte vs. word addresses).
- Clock frequency output for devices w/o own clock - selectable frequency 1 kHz .. 8 MHz
- ISP clock period selection from 0.5 .. 127.5 µs

Optional HW features:
- 4 status LEDs (Prgramming, Read, Write, Error) driven by programmer FW, no avrdude intervention needed.
- Heartbeat LED as ready indicator

Writing random hex files to flash and EEPROM of ATmega328p that had (different) random content before.

Flash write 32K | Flash verify 32K | EEPROM write 1K | EEPROM verify 1K
--|--|--|--
 5.58 s | 4.36 s | 2.82 s | 1.79 s

**Comment by mcuee** [comment]
I do not know any Uno/Nano based PDI programmer for xmega AVR parts.

But if you have ATmega32U4 based Leonardo or Pro Micro boards, maybe you can try out LUFA AVRISP mkII compatible FW to gain PDI capability.
http://fourwalledcubicle.com/blog/2009/12/open-source-xmega-pdi-programmer/
https://github.com/abcminiuser/lufa/tree/master/Projects/AVRISP-MKII

**Comment by mcuee** [comment]
Not using a Uno/Nano but using USB-TTL converter

1) serialupdi PDI programmer
https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md

2) DebugWire programmer
https://github.com/nerdralph/dwprog

3) specific to some FTDI FT232-RL USB-TTL adapter -- SPI or TPI programmer
Example: https://learn.adafruit.com/ftdi-friend/programming-blank-avrs

**Comment by mcuee** [comment]
BTW, if you use `Arduino as ISP` as your main programmer, I will highly recommend buying an AVR ISP Shield from online stores (pretty cheap from places like AliExpress). 

Before I got the shield I used ribbon cables and  I often had connection issues. After I got the shield, it becomes much more stable. 

One example of the AVR ISP Shield. There are other versions.
![Screenshot 2023-04-07 154516](https://user-images.githubusercontent.com/1294740/230566252-c953dcb9-d045-4a87-b9d0-d7c5f98c6363.png)

**Comment by mcuee** [comment]
Side note:

Using the following repo, you can utilize the ATmega16U2 on the Arduino Uno or Mega with the ATmega16U2 as the USB to Serial chip (original or some clones, most of the clones use CH340G or similar chip).
https://github.com/NicoHood/HoodLoader2

You can access both the ATmega16U2 and ATmega328P using avrdude.
```
PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude -patmega16u2 -cavr109 -PCOM35 -b57600
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9489 (probably m16u2)

avrdude done.  Thank you.

PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude -patmega328p -curclock -PCOM35
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e950f (probably m328p)

avrdude done.  Thank you.
```

**Comment by mcuee** [comment]
HVSP/HVPP (other than official tools like stk500v1, stk500v2 and AVR Dragon)

1) ScratchMonkey (STK500v2 based)
https://github.com/microtherion/ScratchMonkey
https://prominimicros.com/how-to-make-an-avr-programmer-out-of-a-pro-mini-part-2-hvsp-hvpp/

2) Atmega32U4 based (eg: using Arduino Leonardo)
https://github.com/MrBlinky/avr-hvsp

3) hvprog and hvprog2
https://mewpro.cc/en/category/hvprog2/

4) not using an Arduino but just parallel port (not for Windows/macOS)
https://github.com/avrdudes/avrdude/pull/1445
